packages/@expo/cli/src/install/fixPackages.ts

```diff
@@ -0,0 +1,92 @@
+import * as PackageManager from '@expo/package-manager';
+import chalk from 'chalk';
+import { spawn } from 'child_process';
+
+import { applyPluginsAsync } from './applyPlugins';
+import * as Log from '../log';
+import { getOperationLog } from '../start/doctor/dependencies/getVersionedPackages';
+import { getVersionedDependenciesAsync } from '../start/doctor/dependencies/validateDependenciesVersions';
+import { groupBy } from '../utils/array';
+
+/**
+ * Given a list of incompatible packages, installs the correct versions of the packages with the package manager used for the project.
+ */
+export async function fixPackagesAsync(
+  projectRoot: string,
+  {
+    packages,
+    packageManager,
+    sdkVersion,
+    packageManagerArguments,
+  }: {
+    packages: Awaited<ReturnType<typeof getVersionedDependenciesAsync>>;
+    /** Package manager to use when installing the versioned packages. */
+    packageManager: PackageManager.NodePackageManager;
+    /**
+     * SDK to version `packages` for.
+     * @example '44.0.0'
+     */
+    sdkVersion: string;
+    /**
+     * Extra parameters to pass to the `packageManager` when installing versioned packages.
+     * @example ['--no-save']
+     */
+    packageManagerArguments: string[];
+  }
+): Promise<void> {
+  if (!packages.length) {
+    return;
+  }
+
+  const { dependencies = [], devDependencies = [] } = groupBy(packages, (dep) => dep.packageType);
+  const versioningMessages = getOperationLog({
+    othersCount: 0, // All fixable packages are versioned
+    nativeModulesCount: packages.length,
+    sdkVersion,
+  });
+
+  const expoDep = dependencies.find((dep) => dep.packageName === 'expo');
+  if (expoDep) {
+    Log.log(
+      chalk`\u203A Updating expo using {bold ${packageManager.name}} and then running {bold npx expo install --fix} under the updated expo version.`
```

Comment:
```suggestion
      chalk`\u203A Updating "expo" package using {bold ${packageManager.name}} and then running {bold npx expo install --fix} under the updated expo version.`
```

i don't think we need to show a message like this, we can just print the actual commands separately. so:

```
npm install expo@^49.0.7
npm install expo-xyz@1.2.3 expo-abc@1.2.4
```

Reply:
Interesting related question as to how the upgrade should actually work. Should it run `npm install expo@version; npm install otherpackage1@version otherpackage2@version` (just outside the current process), or should it install expo first and then run fix/ install again with the remaining packages now that it's on the latest version of expo (e.g., `npm install expo@version; npx expo install --fix`)?

The output you reference comes from `PackageManager` just prior to running the commands... which I'm not currently running :-). I'm doing it the latter way, where I update `expo` first and then run the original command to fix the remaining packages. IMO this is safer for ensuring that everything install/ fix does is done, namely applying config plugins after install. It also would pick up any changes to the install/ fix code itself before installing other packages. In short, it's the closest approximation `npm install expo@latest; npx expo install --fix`, with the possible tradeoff that I'm missing some special options specified in the original command when I append the install/ fix command after the expo install.

In keeping with hopefully at least the spirit of your recommendation, I took away the other warnings and now output exactly the commands that will be run, e.g.,
![image](https://github.com/expo/expo/assets/8053974/b084d1af-35e6-446a-b73f-8abc580ed4d5)

Doing it the other way would actually clean up the `PackageManager` code some, and it would be easy to run exactly what `PackageManager` thinks should be run (just modify `PackageManager` to be able to output what it's running without actually running it). But I'm not sure it's a complete upgrade due to the aforementioned things that wouldn't happen if I did that.

Reply:
1. we probably need to update the "Some dependencies are incompatible..." message:

> Some dependencies are incompatible with the installed expo version:

> The following packages should be updated for best compatibility with Expo SDK 49:

reasoning:

- it's not necessarily true that the packages are "incompatible" - the "expo" package version for example is perfectly fine, it just won't have all of the fixes that we want. sam goes for other packages listed in that list. in other cases a package will be clearly incompatible, but we don't differentiate between those in the message.
- saying that the installed expo version is not compatible with the installed expo version reads quite strangely
 
2. we shouldn't print the "Fix with: npx expo-install --fix` message when you're running this command already

3. if we could handle installing a new version of the expo package in package-manager that would be ideal, but i'm not totally clear on the constraints here given that package-manager itself is a dependency of the expo package. i'd just want to make sure that if we aren't using that directly, then we write the code in a way where we use the correct package manager, and such that this code automatically just works any time we add another package manager (eg: bun)